wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,1265 @@
+    return numerator / denominator
+}
+```
+
+**Test Case:**
+
+```
+@Test
+fun testDivideNumbers_forZeroDenominator_throwsIllegalArgumentException() {
+    val exception = assertThrows<IllegalArgumentException> {
+        divideNumbers(10, 0)
+    }
+    assertThat(exception).contains("Denominator cannot be zero")
+}
+```
+
+Testing exceptions and error handling is vital to ensure that applications behave correctly under error conditions, provide meaningful feedback, and maintain reliability. Without these tests, applications are prone to unpredictable failures, poor user experiences, and potential security issues.
+
+### Verifying Correct Exception Types
+
+Ensure that the correct type of exception is thrown in response to error conditions. This confirms that your error handling logic is specific and accurate.
+
+```
+@Test
+fun testDivideNumbers_forZeroDenominator_throwsIllegalArgumentException() {
+    assertThrows<IllegalArgumentException> {
+        divideNumbers(10, 0)
+    }
+}
+```
+
+### Checking for Proper Exception Messages
+
+It's crucial that exceptions contain meaningful messages that help diagnose issues. Verify that the exception messages are informative and relevant.
+
+```
+@Test
+fun testDivideNumbers_forZeroDenominator_containsCorrectMessage() {
+    val exception = assertThrows<IllegalArgumentException> {
+        divideNumbers(10, 0)
+    }
+    assertThat(exception.message).contains("Denominator cannot be zero")
+}
+```
+
+### Ensuring Exceptions Are Thrown at Correct Times
+
+Exceptions should be thrown only under specific conditions, not during normal operations. Verify that exceptions are correctly managed according to the context.
+
+```
+@Test
+fun testDivideNumbers_forValidInputs_returnsExpectedResult() {
+    val result = divideNumbers(10, 2)
+    assertThat(result).isEqualTo(5)
+}
+```
+
+### Testing Edge Cases
+
+Test edge cases where exceptions might be thrown, such as boundary values or extreme input scenarios.
+
+**Function with Edge Case Handling:**
+
+```
+fun calculateDiscount(price: Double, discountPercent: Double): Double {
+    if (price < 0 || discountPercent < 0) { 
+        throw IllegalArgumentException("Price and discount cannot be negative")
+    }
+    return price - (price * discountPercent / 100)
+}
+```
+
+**Test Case:**
+
+```
+@Test
+fun testCalculateDiscount_forNegativePrice_throwsIllegalArgumentException() {
+    val exception = assertThrows<IllegalArgumentException> {
+        calculateDiscount(-100.0, 10.0)
+    }
+    assertThat(exception.message).contains("Price and discount cannot be negative")
+}
+
+@Test
+fun testCalculateDiscount_forNegativeDiscount_throwsIllegalArgumentException() {
+    val exception = assertThrows<IllegalArgumentException> {
+        calculateDiscount(100.0, -10.0)
+    }
+    assertThat(exception.message).contains("Price and discount cannot be negative")
+}
+```
+
+# Guidelines for Testing Public API
```

Comment:
```suggestion
# Guidelines for Testing Public APIs
```

Content-wise: I think this section is currently serving as a collection of useful examples to demonstrate different aspects of testing public APIs, but it doesn't quite explain the process of _how_ to, which I think is especially useful to attempt since then we're basically teaching people the concepts and methodologies they need in order to incorporate these lessons.

From my perspective, I like to analyze the public API of a class into a list of behaviors that I'll write tests for (I actually usually write out these behaviors in the form of test names using the format you mentioned above, but that's just a personal preference). I _only_ look at this list because, using the test format above, I can very easily see different edge cases that may need to be considered or that I may have missed by looking at the list of behaviors I've written holistically and cross-referencing them with the public API that I'm analyzing.

I don't even start writing tests until this list of behaviors is complete.

I'm wondering if being strongly prescriptive on this approach would be more helpful to readers, especially if they don't yet have a methodology for writing tests based on a public API?

Reply:
Ah, thanks for explaining your approach to writing these tests; that’s super helpful to know in general and for organizing the section. I’ve added a concise version based on how you laid it out. Would like to know if there are any additional steps or details that might be needed. - https://github.com/oppia/oppia-android/blob/b1a5bd76878cfe21ce66ca01df984d0d904c4337/wiki/Writing-tests-with-good-behavioural-coverage.md#testing-public-api